docs: Streamlined the new shield guide (#2515)
zmkfirmware/zmk: docs/docs/development/hardware-integration/new-shield.mdx

};
```

</TabItem>
<TabItem value="split">
For split keyboards you should define your kscan in `my_keyboard.dtsi`.
If your `row-gpios` or your `col-gpios` (or both) are identical between the parts, then they should also be defined in `my_keyboard.dtsi`.

> For example, for a `col2row` 2-part split keyboard (18 keys split into a 3x3 macropad on both halves) where the "row" GPIOs used are the same for both  halves:

Reviewer: ```suggestion
For example, for a `col2row` 2-part split keyboard (18 keys split into a 3x3 macropad on both halves) where the "row" GPIOs used are the same for both halves:
```

```dts title="my_keyboard.dtsi"
/ {
    kscan0: kscan0 {
        compatible = "zmk,kscan-gpio-matrix";
        diode-direction = "col2row";
        wakeup-source;

        row-gpios
            = <&pro_micro 6 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>
            , <&pro_micro 7 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>
            , <&pro_micro 8 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>
            ;

    };
};
```

The missing `col-gpios` would be defined in your `my_keyboard_left.overlay` and `my_keyboard_right.overlay` files.

```dts title="my_keyboard_left.overlay"

> #include "my_keyboard.dtsi" // The main dtsi files are included in the overlay.

Reviewer: ```suggestion
#include "my_keyboard.dtsi" // The shared dtsi file is included in the overlay
```

// Label of the kscan node in the dtsi
&kscan0 {
    col-gpios
        = <&pro_micro 19 GPIO_ACTIVE_HIGH>
        , <&pro_micro 18 GPIO_ACTIVE_HIGH>
        , <&pro_micro 15 GPIO_ACTIVE_HIGH>
        ;
};
```

```dts title="my_keyboard_right.overlay"

> #include "my_keyboard.dtsi" // The main dtsi files are included in the overlay.

Reviewer: ```suggestion
#include "my_keyboard.dtsi" // The shared dtsi file is included in the overlay
```

// Label of the kscan node in the dtsi
&kscan0 {
    col-gpios
        = <&pro_micro 10 GPIO_ACTIVE_HIGH>
        , <&pro_micro 11 GPIO_ACTIVE_HIGH>
        , <&pro_micro 13 GPIO_ACTIVE_HIGH>
        ;
};
```

</TabItem>
</SplitTabs>

### Matrix Transform

The matrix transform is used to transform row/column events into "key position" events.

> When a key is pressed, a kscan event is generated from it with a row value and a column value corresponding to the index (from 0) of the `row-gpios` pin and `col-gpios` pin respectively that triggered the event.

Reviewer: ```suggestion
When a key is pressed, a kscan event is generated from it with a `row` and a `column` value corresponding to the zero-based indices of the `row-gpios` and `col-gpios` pins that triggered the event, respectively.
```
I think the sentence didn't flow well; this seems an improvement to me although it isn't perfect.

> The "key position" triggered is the index of the `RC(row, column)` in the matrix transform that correponds to the kscan event.

Reviewer: ```suggestion
Then, the "key position" triggered is the index of the `RC(row, column)` in the matrix transform where `row` and `column` are the indices as mentioned above.
This key position will in turn have a behavior binding associated with it in the keymap.
```

<SplitTabs>
<TabItem value="unibody">
The `my_keyboard.overlay` must include a matrix transform that defines this mapping from row/column values to key positions.
Add `#include <dt-bindings/zmk/matrix_transform.h>` to the top of the file.

> Here is an example of a matrix transform for the previous 3x3 macropad:

Reviewer: I like the simpler example here, although I wonder if it should at least have a missing entry or something of the sort. If you can think of such a simple example that can replace this easily it would be nice to have.

Author: I played around with it a bit, but I think a more complex example also needs to be larger in order to be intuitive, and that scales up the page significantly. I added a line pointing to the in-tree examples instead.

:::

</TabItem>
</Tabs>
When a key is pressed, a kscan event is generated from it with a row value and a column value corresponding to the index (from 0) of the `row-gpios` pin and `col-gpios` pin respectively that triggered the event.
The "key position" triggered is the index of the `RC(row, column)` in the matrix transform that correponds to the kscan event.

## Matrix Transform
<SplitTabs>
<TabItem value="unibody">
The `my_keyboard.overlay` must include a matrix transform that defines this mapping from row/column values to key positions.
Add `#include <dt-bindings/zmk/matrix_transform.h>` to the top of the file.

Internally ZMK translates all row/column events into "key position" events to maintain a consistent model that works no matter what any possible GPIO matrix may look like for a certain keyboard. This is particularly helpful when:
Here is an example of a matrix transform for the previous 3x3 macropad:

1. To reduce the used pins, an "efficient" number of rows/columns for the GPIO matrix is used, that does _not_ match the physical layout of rows/columns of the actual key switches.

> 1. For non rectangular keyboards with thumb clusters, non `1u` locations, etc.

Reviewer: While your new explanation seems good, it is pretty concise and I think it would be good to keep some of the explanations here, like these scenarios. When we have the dedicated explainer page we can remove some of the details.

Author: Added these back on the bottom of the page.

Reviewer: Thanks, the new remarks look good.

> To allow the kscan matricies to be joined in the matrix transform, an offset is applied to the matrix transform of peripherals.

Reviewer: ```suggestion
To allow the kscan matrices to be joined in the matrix transform, an offset is applied to the matrix transform of peripherals.
```

> This offset means that when the right half of the keyboard has a keypress event triggered by the GPIO pins at the indicies `0,0` of its `row-gpios` and `col-gpios` arrays respectively, it will interpret it as an `RC(0,3)` event rather than an `RC(0,0)` event.

Reviewer: ```suggestion
This offset means that when the right half of the keyboard has a key event triggered by the GPIO pins at the indices `0,0` of its `row-gpios` and `col-gpios` arrays respectively, it will interpret it as an `RC(0,3)` event rather than an `RC(0,0)` event.
```

To allow the kscan matrices to be joined in the matrix transform, an offset is applied to the matrix transform of peripherals.

```dts title="my_keyboard_right.overlay"
&default_transform { // Offset of 3 because the left side has 3 columns
    col-offset = <3>;
};
```

This offset means that when the right half of the keyboard has a key event triggered by the GPIO pins at the indices `0,0` of its `row-gpios` and `col-gpios` arrays respectively, it will interpret it as an `RC(0,3)` event rather than an `RC(0,0)` event.
Additional peripherals would need their columns to be offset by an ever increasing number equal to the sum of the columns in the central + any peripherals that came before it.
You can also apply row offsets with `row-offset`.

</TabItem>
</SplitTabs>

The matrix transform is also used to "correct" pin orderings into something that more closely matches the physical order of keys. Causes of abnormal pin orderings include:

- To reduce the used pins, an "efficient" number of rows/columns for the GPIO matrix is used, that does _not_ match the physical layout of rows/columns of the actual key switches.

> - For non rectangular keyboards with thumb clusters, non `1u` locations, etc.

Reviewer: ```suggestion
- For non-rectangular keyboards with thumb clusters, non `1u` locations, etc.
```

};
```

> See [Studio Setup](./studio-setup.md) for information on defining the physical key positions enabling [ZMK Studio](../../features/studio.md) for your keyboard.

Reviewer: ```suggestion
See [ZMK Studio Setup](./studio-setup.md) for information on defining the physical key positions property that lets you enable [ZMK Studio](../../features/studio.md) for your keyboard.
```

### Chosen Node

> Set the `chosen` node to a defined "default" physical layout.

Reviewer: ```suggestion
Set the `chosen` node to a defined "default" physical layout. This should also be placed in the same file as the physical layout, i.e. `my_keyboard.overlay` for unibodies and `my_keyboard.dtsi` for split keyboards.
```

```dts
/ {
    chosen {
        zmk,kscan = &kscan0;
        zmk,physical-layout = &default_layout;
        // Other chosen items
    };
};
```

:::

## Default Keymap

Each keyboard should provide a default keymap to be used when building the firmware, which can be overridden and customized by user configs.
For "shield keyboards", this should be placed in the `boards/shields/my_keyboard/my_keyboard.keymap` file.
The keymap is configured as an additional devicetree overlay that includes the following:

Here is an example simple keymap for a 3x3 macropad, with only one layer:

```dts title="my_keyboard.keymap"
/ {
    keymap {
        compatible = "zmk,keymap";

        default_layer { // Layer 0
            // -------------------------------------
            // |     Z     |     M     |     K     |
            // |     A     |     B     |     C     |
            // |     D     |     E     |     F     |
            bindings = <
                &kp Z    &kp M    &kp K
                &kp A    &kp B    &kp C
                &kp D    &kp E    &kp F
            >;
        };
    };
 };
```

See [Keymaps](../../keymaps/index.mdx) for information on defining keymaps in ZMK.

> ## Metadata

Reviewer: I wonder if this should be taken out and replaced with a pointer as well, since the same example exists in the dedicated page and this section doesn't add much.
